### PR TITLE
Specify the hash algorithm in the sc library functions #610

### PR DESCRIPTION
**Pull request by apoorv-2204**
# Description
- This pr Updates the existing function in SC library `hash` to support multiple hash algos.
``` elixir
      hash("data","sha256")
      hash("data","sha512")   
      hash("data","sha3_256")
      hash("data","sha3_512")
      hash("data","blake2b")
```
- Also introduces an unused aehash function for ae padding

- [x] #610 

## Type of change

Please delete options that are not relevant.

- New feature (non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing functionality to not work as expected)

# How Has This Been Tested?

- Unit Tests

# Checklist:

- My code follows the style guidelines of this project
- I have performed a self-review of my own code
- I have commented my code, particularly in hard-to-understand areas
- I have made corresponding changes to the documentation
- My changes generate no new warnings
- I have added tests that prove my fix is effective or that my feature works
- New and existing unit tests pass locally with my changes
- Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: lib/archethic/contracts/interpreter/utils.ex [hunk 1]
```diff
@@ -216,7 +216,7 @@ defmodule Archethic.Contracts.Interpreter.Utils do
   end
 
   # Whitelist the hash/1 function
-  def prewalk(node = {{:atom, "hash"}, _, [_data]}, acc = {:ok, %{scope: scope}})
+  def prewalk(node = {{:atom, "hash"}, _, [_data, _algo]}, acc = {:ok, %{scope: scope}})
```
Comment (samuelmanzanera) [inline lib/archethic/contracts/interpreter/utils.ex:219]: I think you could provide a check to make the sure the algo is supported

File: lib/archethic/contracts/interpreter/library.ex [hunk 2]
```diff
@@ -108,12 +109,68 @@ defmodule Archethic.Contracts.Interpreter.Library do
...
+        "blake2b" ->
+          :blake2b
+      end
+
+    :crypto.hash(algo, decode_binary(content))
+    |> Base.encode16()
+  end
+
+  def list_supported_hash(),
+    do: ["sha256", "sha512", "sha3_256", "sha3_512", "blake2b"]
```
Comment (samuelmanzanera) [inline lib/archethic/contracts/interpreter/library.ex:144]: I think you can reuse the already existing list on the config

File: lib/archethic/contracts/interpreter/library.ex [hunk 3]
```diff
@@ -108,12 +109,65 @@ defmodule Archethic.Contracts.Interpreter.Library do
...
+    :crypto.hash(algo, decode_binary(content))
+    |> Base.encode16()
+  end
+
+  @spec ae_hash(
+          content :: binary(),
+          algo :: binary()
+        ) ::
+          binary()
+  def ae_hash(content, algo) do
```
Comment (samuelmanzanera) [inline lib/archethic/contracts/interpreter/library.ex:148]: Why do you need this function?

File: lib/archethic/contracts/interpreter/utils.ex [hunk 4]
```diff
@@ -215,9 +217,9 @@ defmodule Archethic.Contracts.Interpreter.Utils do
     {node, acc}
   end
 
-  # Whitelist the hash/1 function
-  def prewalk(node = {{:atom, "hash"}, _, [_data]}, acc = {:ok, %{scope: scope}})
-      when scope != :root,
+  # Whitelist the hash/2 function
+  def prewalk(node = {{:atom, "hash"}, _, [_data, algo]}, acc = {:ok, %{scope: scope}})
```
Comment (samuelmanzanera) [inline lib/archethic/contracts/interpreter/utils.ex:221]: I think you should keep `hash/1` to be backward compatible and use the default hash algorithm in the library function

Comment (apoorv-2204) [reply lib/archethic/contracts/interpreter/utils.ex:221]: done

**Review by Neylix — APPROVED** [review]

**Pull request merged**